Adds API in semanticsconfiguration to decide how to merge child semanticsConfigurations

chunhtai opened the pull request:
fixes https://github.com/flutter/flutter/issues/110158

The goal of this pr is to let _RenderDecoration to be able to choose which children to merge together without merging upward. 

The requirements are:

1. A render object can decide which children to merge and form a node.
2. a render object must be able to control the order of the merged nodes, including the one the render object itself formed
3. TextField must remain mergeable if wrapped with a semantics widget

The new API SemanticsConfiguration.merger let renderobject to decide which of its children can merge together. The formed semantics node will attach to the **parent** of the semantics node the renderobject created or merged to, so they will be siblings.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

chunhtai commented:
Hi @goderbauer and @jonahwilliams  this is still work in progress, but I would like to get your feedback first. Thanks in advance!

jonahwilliams reviewed (commented):
Can you give a high level overview of how this should be used?

chunhtai commented:
A recap on how semantics tree is formed.
1. starts with rendering tree, it creates _SemanticsFragments for each render object, and calls describeSemanticsConfiguration to get the SemanticsConfiguration from each render object and attaches to its _SemanticsFragments
2. For each node, based on SemanticsConfiguration, it can either form a semanticsnode and merge all the SemanticsConfigurations from child _SemanticsFragments if possible, or it sends all the _SemanticsFragments up one level to let parent decide what to do with it.

All these behavior are more or less hard coded or inferred by the properties in SemanticsConfigurations.

This PR adds a callback in 2. Before it try to do anything with the child SemanticsConfigurations, it calls the callback to do whatever it wants and then proceed with the regular merge.

What this enable us is that a rendering object can decide how to merge children, or even make a child to form its own semantics node.

goderbauer commented:
Looks like the checks are unhappy.

goderbauer commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -1316,6 +1316,35 @@ class _RenderDecoration extends RenderBox with SlottedContainerRenderObjectMixin
+
+  @override
+  void describeSemanticsConfiguration(SemanticsConfiguration config) {
+    config.childSemanticsConfigMutator = (Map<AbstractNode, List<SemanticsConfiguration>> childConfigs) {
```
Would something like this work: a callback that gives you access to your render children and the associated SemanticConfig and you return the configs that you want to have merged into the parent. The remaining one stay explicit.

chunhtai commented on packages/flutter/lib/src/material/date_picker.dart:
```diff
@@ -769,7 +769,15 @@ class _DatePickerHeader extends StatelessWidget {
     );
-
+    final Widget? effectiveEntryModeButton;
+    if (entryModeButton != null) {
```
This is previous not needed because the `entryModeButton` conflicted with TextField and marked as explicit. After the patch the TextField creates an empty SemanticsConfig with a bunch of explicit child semantics node. so the semantics config no longer conflict and the icon button gets merge to parent and create weird semantics.

This fix makes sure it won't ever merge up

chunhtai committed:
Adds semantics merger API and fix input decorator

chunhtai commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -3129,7 +3129,12 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
     final bool wasSemanticsBoundary = _semantics != null && (_cachedSemanticsConfiguration?.isSemanticBoundary ?? false);
     _cachedSemanticsConfiguration = null;
-    bool isEffectiveSemanticsBoundary = _semanticsConfiguration.isSemanticBoundary && wasSemanticsBoundary;
+    // Merger may produce sibling node to be attached to the parent of this
```
This is non-ideal if the merger does not produce a sibling merge group. The text widget will likely need to use the merger to fix https://github.com/flutter/flutter/issues/111931. This means if the text has widget span, everytime the text widget is mark semantics dirty, it will need to go up two level and marks the grandparent dirty. Not sure if this is a big deal.

If we want to optimize this, we need two APIs, one for merge up rule one for sibling merge rule. A semanticsConfiguration only need to marks grandparent dirty if it has sibling merge rule.

chunhtai replied:
the pr to fix text widget with merger is https://github.com/flutter/flutter/pull/113461

goderbauer replied:
> If we want to optimize this, we need two APIs, one for merge up rule one for sibling merge rule. A semanticsConfiguration only need to marks grandparent dirty if it has sibling merge rule.

For my own understanding: Is this a performance issue? Or is there also a correctness issue here?

chunhtai replied:
there is also a correctness issue. 

If a renderobject has a sibling merge group, the node created by the sibling merge group will be the sibling of the semantics node this renderobject creates. If this renderobject is not explicit, then the sibling group will look up to find the first explicit rendering parent and be its sibling.

goderbauer replied:
> there is also a correctness issue.

If there's also a correctness issue, that's not great. Can we disallow the situation that causes the correctness issue?

> If a renderobject has a sibling merge group, the node created by the sibling merge group will be the sibling of the semantics node this renderobject creates. If this renderobject is not explicit, then the sibling group will look up to find the first explicit rendering parent and be its sibling.

Actually, what is the correctness issue here? Isn't this exactly what we want to achieve with this? The siblingmergegroups will become a sibling of the node that the other configurations were merged into?

chunhtai replied:
oh sorry I meant something will break if it doesn't go above one more level to mark renderobject semantics dirty

GaryQian commented:
@goderbauer This seems like it is ready for another round of review.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -48,6 +51,28 @@ typedef SetTextHandler = void Function(String text);
 
+/// The semantics merge rule generated from [SemanticsConfiguration.merger].
+class SemanticsMergeRule {
+  /// The child semantics configuration that are merging upward.
```
Are the always merged up? What if they conflict with semantics information of the parent?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -48,6 +51,28 @@ typedef SetTextHandler = void Function(String text);
+  /// The groups of child semantics configuration that want to merge together
+  /// and form a child [SemanticsNode].
+  ///
+  /// If the [SemanticsConfiguration] in the group is a semantics boundary or
```
This is a list of semanticsCondfigurations. Which one is "the" one that determines this?

goderbauer replied:
Oh, do you mean _all_ the semanticsConfigurations that are boundaries or conflict will be excluded from the merge? Maybe say that: All the [SemanticsConfiguration]s in a given group that are either semantics boundaries or are conflicting with other [SemanticsConfiguration]s of the same group will be excluded from the sibling merge group and form independent semantics nodes as usual.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4422,6 +4447,20 @@ class SemanticsConfiguration {
+  ///
+  /// The input is a lookup map that maps rendering children to their semantics
+  /// configurations. The map can be empty if the rendering object of this
+  /// semantics configuration a leaf node.
```
nit: the last sentence appears to be incomplete.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -48,6 +51,28 @@ typedef SetTextHandler = void Function(String text);
 typedef SemanticsActionHandler = void Function(Object? args);
 
+/// The semantics merge rule generated from [SemanticsConfiguration.merger].
+class SemanticsMergeRule {
```
The name "Rule" is a little odd. It's not really defining a rule, isn't it more the result of calling the merger?

goderbauer replied:
I would have expected a constructor here where you have to pass in the two lists. I don't think we should encourage people to manipulate the lists directly.

goderbauer replied:
`@immutable`?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4422,6 +4447,20 @@ class SemanticsConfiguration {
   }
 
+  /// A function that decide how configuration should merge with each other.
+  ///
```
Maybe document somewhere (either here or on the SemanticsMergeRule class or probably on both) what exactly the merger gets to decide: e.g.

The merger function gets the SemanticsConfigurations of the direct descendent nodes as input and can decide which of these should be merged together to form child SemanticsNodes and which of them should be merged upwards into the parent SemanticsNode.

goderbauer commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -1314,6 +1314,29 @@ class _RenderDecoration extends RenderBox with SlottedContainerRenderObjectMixin
+      if (prefix != null) {
+        final List<SemanticsConfiguration>? prefixMergeGroup = childConfigs
+            .remove(prefix);
+        if (prefixMergeGroup != null && prefixMergeGroup.isNotEmpty) {
```
this seems odd, why do you have to check for both? is there any meaning behind having an empty list in `childConfigs`?

If so, this should be documented on the typedef for merger and maybe on the merger property as well.

goderbauer commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -1314,6 +1314,29 @@ class _RenderDecoration extends RenderBox with SlottedContainerRenderObjectMixin
+      final SemanticsMergeRule rule = SemanticsMergeRule();
+      if (prefix != null) {
+        final List<SemanticsConfiguration>? prefixMergeGroup = childConfigs
+            .remove(prefix);
```
formatting nit: just leave this on one line. (same below)

goderbauer commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -3129,7 +3129,12 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
     final bool wasSemanticsBoundary = _semantics != null && (_cachedSemanticsConfiguration?.isSemanticBoundary ?? false);
     _cachedSemanticsConfiguration = null;
-    bool isEffectiveSemanticsBoundary = _semanticsConfiguration.isSemanticBoundary && wasSemanticsBoundary;
+    // Merger may produce sibling node to be attached to the parent of this
```
Should we instead just require explicitly that if you want to specify "merger" you have to be a semanticsBoundary as well? Why do we have to do this implicitly?

chunhtai replied:
the problem is the sibling merge group will mutate the children of the semanticsboundary above the immediate semantics boundary, so even if the current renderobject is a semantics boundary, it still need to go up one level and mark the parent semantics boundary dirty.

chunhtai replied:
and we can't require explcity if there is a merger so that the renderobject still has the choose to merge up

goderbauer commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -3203,70 +3210,97 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
+    assert(hasSemanticsMerger || configToFragment.isEmpty);
+
+    if (explicitChildNode) {
+      assert(!hasSemanticsMerger, 'A Semantics with `explicitChildNode = true` cannot have a merger');
```
Can we enforce this right when you configure the SemanticsConfigurations object instead of producing an error during semantics compilation?

chunhtai replied:
sounds good

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4512,8 +4551,9 @@ class SemanticsConfiguration {
   /// absorb other configurations and it is recommended to only absorb compatible
   /// configurations as determined by [isCompatibleWith].
   void absorb(SemanticsConfiguration child) {
+    // TODO(chunhtai): this should assert more before absorb.
```
this is old code i forgot to remove. will remove it

chunhtai committed:
addressing comments

chunhtai commented:
@goderbauer Thanks for the review, This is ready for another look. google testing failure is expected a11y scuba change

jonahwilliams commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -29,6 +29,15 @@ export 'semantics_event.dart' show SemanticsEvent;
+/// Signature for the [SemanticsConfiguration.merger].
+///
+/// The input map can be used for looking up [SemanticsConfiguration]s rendering
+/// children produces. The keys are [RenderObject]s, and the values are their
```
If the keys are always RenderObjects, I would type this as `Map<RenderObject, List<SemanticsConfiguration>>`, unless you don't want folks to access the render object, in which case I would use `Map<Object, List<SemanticsConfiguration>>`

We've definitely talked about removing the AbstractNode usage before, so I would avoid exposing it via public API.

chunhtai replied:
This is in semantics layer so I don't have access to renderobject class that is why I used AbstractNode, I will use `Object` instead

goderbauer replied:
Elsewhere we use SemanticsTag to identify SemanticsNodes coming from a particular RenderObject because RenderObjects are not a concept the semantics layer knows about. I wonder if we could reuse that concept here for the configurations as well instead of awkwardly mapping the configurations to an `Object` of unknown type because the type we really want here doesn't exist in this layer.

chunhtai committed:
abstractnode to object

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -53,6 +62,44 @@ typedef SemanticsActionHandler = void Function(Object? args);
+  /// nodes as usual.
+  final List<SemanticsConfiguration> mergeUp;
+
+  /// The groups of child semantics configuration that want to merge together
```
configuration -> configurations.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4438,6 +4485,25 @@ class SemanticsConfiguration {
     _hasBeenAnnotated = true;
   }
 
+  /// A function that decide how configuration should merge with each other.
```
configuration -> configurations of descendants

goderbauer commented on packages/flutter/test/material/text_field_test.dart:
```diff
@@ -4268,6 +4268,47 @@ void main() {
+
   testWidgets('TextField with specified suffixStyle', (WidgetTester tester) async {
     final TextStyle suffixStyle = TextStyle(
       color: Colors.pink[500],
```
There should probably be some tests that test the merger functionality independent of the Textfield in various configurations to ensure it has the expected results.

chunhtai replied:
added more tests

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -29,6 +29,15 @@ export 'semantics_event.dart' show SemanticsEvent;
+/// [SemanticsConfiguration]s. The list of each value is always non-empty.
+///
+/// The return value is a [SemanticsMergeResult].
+typedef SemanticsConfigMerger = SemanticsMergeResult Function(Map<Object, List<SemanticsConfiguration>>);
```
Do you always have to return all SemnaticsConfigurations from the map in the SemanticsMergeResult? What does it mean if you drop some of them on the floor? Is that a bug we should guard against?

Or should this really be just returning a `List<List<SemanticsConfiguration>>`, the siblings merge groups. And all SemnaticsConfigurations not included here are merged up as usual?

goderbauer replied:
The documented behavior would then be:

The SemanticsConfigMerger decides which SemanticsConfigurations of descendants should form explicit nodes. SemanticsConfigurations that are not made into explicit nodes are merged upwards and the explicit nodes are made into siblings of the upwards merge target.

goderbauer replied:
The fact that these explicit nodes are siblings and not children of the parent is really odd now that I see it written out explicitly. Are we sure we are doing the right thing here?

chunhtai replied:
> Do you always have to return all SemnaticsConfigurations from the map in the SemanticsMergeResult? What does it mean if you drop some of them on the floor? Is that a bug we should guard against?

I think dropping some configruations are ok. There will be use case where we want to insert additional synthetic configurations though.

If we return `List<List<SemanticsConfiguration>>` we will lose the functionality for both dropping or adding configurations.

> The fact that these explicit nodes are siblings and not children of the parent is really odd now that I see it written out explicitly.

This is for this specific use case, I don't have a use case where we want the merge group to be children of the parent. We can always add more api to SemanticsMergeRule if there is use case for it.

goderbauer reviewed (commented):
Since this is such an obscure and somewhat complicated feature, it needs better documentation and thorough test coverage.

goderbauer commented:
> TextField must remain mergeable if wrapped with a semantics widget

This requirement is actually kinda odd and I think the main cause for the complexity this approach is introducing. Without this, the problem could be fixed with existing features (e.g. semantics boundaries and maybe custom assembleSemanticsNode implementations). Why is this a requirement?

chunhtai commented:
> 
> This requirement is actually kinda odd and I think the main cause for the complexity this approach is introducing. Without this, the problem could be fixed with existing features (e.g. semantics boundaries and maybe custom assembleSemanticsNode implementations). Why is this a requirement?

I saw use cases where people use wraps semantics label around the textfield. or the textfield itself wraps Semantics.onTap around the InputDecorator as well.

Even if we removed this requirement the other two requirements still seems non-trivial to fix with current assembleSemanticsNode

1. A render object can decide which children to merge and form a node.
2. a render object must be able to control the order of the merged nodes, including the one the render object itself formed

chunhtai committed:
feature complete

chunhtai commented:
1. Renamed everything to be in the line of `child, semanticsconfigs, delegate`.
2. Now the delegate takes a list of semanticsConfigurations, and it will be rendering object's job to mark semanticsconfiguration with tag if they want to look up semanticsconfigs for a child
3. Added an API in SemanticsConfig to looks up what semantics tag it has. so that (2) is possible.
4. Added more test and doc

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4448,6 +4571,11 @@ class SemanticsConfiguration {
   ///    usage.
   Iterable<SemanticsTag>? get tagsForChildren => _tagsForChildren;
+
+  /// Whether the this configuration will tag the child semantics nodes with a
```
remove the "the"

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -4448,6 +4571,11 @@ class SemanticsConfiguration {
+
+  /// Whether the this configuration will tag the child semantics nodes with a
+  /// given [SemanticsTag].
+  bool isChildrenTagged(SemanticsTag tag) => _tagsForChildren?.contains(tag) ?? false;
```
maybe rename to: `tagsChildrenWith`, `isChildrenTagged` doesn't sound like correct english.

goderbauer commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -1683,12 +1712,16 @@ class _AffixText extends StatelessWidget {
     this.text,
     this.style,
     this.child,
+    required this.ordinalKey,
```
nit: rename to `sortKey`?

goderbauer commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -3158,7 +3162,12 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
     final bool wasSemanticsBoundary = _semantics != null && (_cachedSemanticsConfiguration?.isSemanticBoundary ?? false);
     _cachedSemanticsConfiguration = null;
-    bool isEffectiveSemanticsBoundary = _semanticsConfiguration.isSemanticBoundary && wasSemanticsBoundary;
+    // Merger may produce sibling node to be attached to the parent of this
```
nit: eliminate the word "Merger" here since that's not what this is called anymore?

chunhtai replied:
ah sorry for this, my IDE lied to me again, I did a search all on keyword merger, but it still not catching this and several instances. Will fix

goderbauer commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -3232,70 +3243,94 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
-
+    final List<SemanticsConfiguration> childConfigurations = <SemanticsConfiguration>[];
+    final bool explicitChildNode = config.explicitChildNodes || parent is! RenderObject;
+    final bool hasSemanticsMerger = config.childConfigsDelegate != null;
```
same here about the word "merger". Can you audit the entire change to make sure we eliminate this name since the concept no longer exists?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -53,6 +54,20 @@ typedef SemanticsActionHandler = void Function(Object? args);
 /// Used by [SemanticsOwner.onSemanticsUpdate].
 typedef SemanticsUpdateCallback = void Function(ui.SemanticsUpdate update);
 
+/// Signature for the [SemanticsConfiguration.childConfigsDelegate].
```
here and elsewhere: In line with our style guide we probably shouldn't be using the "configs" abbreviation and instead call this "configuration" everywhere.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -53,6 +54,20 @@ typedef SemanticsActionHandler = void Function(Object? args);
+/// The return value is the arrangement of these configs, including which
+/// configs continue to merge upward and which configs form sibling merge group.
+///
+/// Consider using [ChildSemanticsConfigsResultBuilder] to generate the return
```
Let's remove the consider and make that the only way?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -85,6 +100,95 @@ class SemanticsTag {
+/// should form.
+///
+/// Consider using [ChildSemanticsConfigsResultBuilder] to build the result.
+abstract class ChildSemanticsConfigsResult {
```
Why split this in an abstract and a private class? Can this just be a concrete class with a private constructor that's only used by the builder?

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -85,6 +100,95 @@ class SemanticsTag {
+///
+/// Consider using [ChildSemanticsConfigsResultBuilder] to build the result.
+abstract class ChildSemanticsConfigsResult {
+  /// Gets [SemanticsConfiguration]s that are merging into parent semantics node.
+  ///
+  /// All merging up [SemanticsConfiguration]s that are either semantics
+  /// boundaries or are conflicting with other [SemanticsConfiguration]s
+  /// in the parent will form independent semantics nodes as usual. All other
+  /// [SemanticsConfiguration]s will be merged into parent semantics node.
```
Returns the [SemanticsConfiguration]s that are supposed to be merged into the parent semantics node.

[SemanticsConfiguration]s that are either semantics boundaries or are conflicting with other [SemanticsConfiguration]s will form explicit semantics nodes. All others will be merged into the parent.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -85,6 +100,95 @@ class SemanticsTag {
+  /// boundaries or are conflicting with other [SemanticsConfiguration]s
+  /// in the parent will form independent semantics nodes as usual. All other
+  /// [SemanticsConfiguration]s will be merged into parent semantics node.
+  Iterable<SemanticsConfiguration> get mergeUp;
```
Here and below, maybe type these as `List`s since that's what they are gonna be anyways.

goderbauer commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -85,6 +100,95 @@ class SemanticsTag {
+  final List<SemanticsConfiguration> _mergeUp = <SemanticsConfiguration>[];
+  final List<List<SemanticsConfiguration>> _siblingMergeGroups = <List<SemanticsConfiguration>>[];
+
+  /// Marks the [SemanticsConfiguration] to be merged into parent semantics
```
into _the_ parent

goderbauer reviewed (commented):
Left some comments, but overall this LGTM

chunhtai committed:
addressing comments

chunhtai commented:
@goderbauer requested changes are in last commit

goderbauer commented on packages/flutter/lib/src/rendering/object.dart:
```diff
@@ -3161,7 +3165,13 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
     final bool wasSemanticsBoundary = _semantics != null && (_cachedSemanticsConfiguration?.isSemanticBoundary ?? false);
     _cachedSemanticsConfiguration = null;
-    bool isEffectiveSemanticsBoundary = _semanticsConfiguration.isSemanticBoundary && wasSemanticsBoundary;
+    // The childConfigurationsDelegate may produce sibling node to be attached
```
may produce _a_ sibling node

 or 
may produce sibling _nodes_

Is it one or many?

chunhtai replied:
many, will fix

goderbauer reviewed (approved):
LGTM

chunhtai committed:
fix comments

chunhtai committed:
conditionally add sort order

chunhtai committed:
fix bool

chunhtai committed:
fix test

chunhtai committed:
more fix

chunhtai committed:
fix tests
